### PR TITLE
[SDK-1279] getTokenSilently retry logic

### PR DESCRIPTION
**Pull request by stevehobbsdev**
### Description

Implements retry logic when using `fetch` to get a new access token. The implementation relies on `fetch` only failing in the event of a network failure. Failure responses, such as 4xx or 5xx errors still return a resolved promise but with the error detail in the body.

### References

[Fetch API on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)

### Testing

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

File: src/constants.ts [hunk 1]
```diff
@@ -9,3 +9,8 @@ export const DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS = 60;
 export const DEFAULT_POPUP_CONFIG_OPTIONS: PopupConfigOptions = {
   timeoutInSeconds: DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS
 };
+
+/**
+ * @ignore
+ */
+export const DEFAULT_SILENT_TOKEN_RETRY_COUNT = 5;
```
Comment (joshcanhelp) [inline src/constants.ts]: This doesn't look like it's configurable ... is it? Shouldn't it be? Also, default 5 times seems like a lot ...

Comment (stevehobbsdev) [reply src/constants.ts]: I started down the road of making it configurable but ultimately it feels like as a first pass it would be better to set this to something and see what feedback we get around it. It might be what we have right now (a sensible default) is enough for 99% of cases.

I'm not precious about the 5, but if you feel it's too high, the only other number I suggest would be 3. WYDT?

Comment (joshcanhelp) [reply src/constants.ts]: As long as you think this has been built in a way to make adding that config easily, then I'm 👍 

Given that the timeout is 60 seconds (😮), 5 retries is potentially ... 6 minutes?? Is the UI frozen for that time? For network failures I feel like 1, maybe 2 is sufficient.

Comment (stevehobbsdev) [reply src/constants.ts]: Ah, so the timeout you're referring to is our own timeout for the `/authorize` call when using an iframe. You raise a good point though, I'm not sure what the default timeout for `fetch` is when it comes to network errors - let me check that out first.

Comment (stevehobbsdev) [reply src/constants.ts]: To circle back on this, `fetch` itself doesn't prescribe any time out at all. So it will essentially wait until something fails before returning.

To combat this, I will wrap the fetch call and the retries in a wrapper than enforces a timeout, say, 10 seconds, and lower the retry count to 3. How does that sound?

Comment (joshcanhelp) [reply src/constants.ts]: Sounds much better!

**Comment by joshcanhelp** [comment]
@stevehobbsdev - Feel free to ping me when this is back in shape!

**Comment by stevehobbsdev** [comment]
@joshcanhelp This is ready for you to look at again please.

File: src/utils.ts [hunk 2]
```diff
@@ -182,25 +188,62 @@ export const bufferToBase64UrlEncoded = input => {
   );
 };
 
-const getJSON = async (url, options) => {
-  const response = await fetch(url, options);
+const fetchWithTimeout = (url, options, timeout = DEFAULT_FETCH_TIMEOUT_MS) => {
+  // The promise will resolve with one of these two promises (the fetch and the timeout), whichever completes first.
+  return Promise.race([
```
Comment (joshcanhelp) [inline src/utils.ts]: `Promise.race` ... handy!

Comment (joshcanhelp) [reply src/utils.ts]: I assume this ends the `fetch()` process so that's not being tried in the background while the app continues?

File: src/utils.ts [hunk 3]
```diff
@@ -182,25 +188,62 @@ export const bufferToBase64UrlEncoded = input => {
...
+        timeout
+      );
+    })
+  ]);
+};
+
+const getJSON = async (url, timeout, options) => {
+  let fetchError, response;
+
+  for (let i = 0; i < DEFAULT_SILENT_TOKEN_RETRY_COUNT; i++) {
```
Comment (lbalmaceda) [inline src/utils.ts]: seems like `DEFAULT_SILENT_TOKEN_RETRY_COUNT` is hardcoded and not customizable by the user. Do we want it that way or should we open an option value here?

File: src/utils.ts [hunk 4]
```diff
@@ -182,25 +188,62 @@ export const bufferToBase64UrlEncoded = input => {
   );
 };
 
-const getJSON = async (url, options) => {
-  const response = await fetch(url, options);
+const fetchWithTimeout = (url, options, timeout = DEFAULT_FETCH_TIMEOUT_MS) => {
+  // The promise will resolve with one of these two promises (the fetch and the timeout), whichever completes first.
+  return Promise.race([
+    fetch(url, options),
+    new Promise((_, reject) => {
```
Comment (lbalmaceda) [inline src/utils.ts]: I'm sure this works for rejecting the promise early and eventually retrying. But to me, it feels like the `timeout` should be a feature/option of the node-fetch library. Because, if timeout kicks in, the request should be correctly canceled and that's only something the client can do.

File: src/utils.ts (hunk 3 above)
Comment (stevehobbsdev) [reply src/utils.ts]: It feels like something we can leave at a sensible default and see what feedback we get around it. I can't see people needing to change this value too much.

File: src/utils.ts (hunk 4 above)
Comment (stevehobbsdev) [reply src/utils.ts]: You are perhaps correct. However, we don't use `node-fetch` here, it's just the browsers Fetch API (polyfilled using [`promise-polyfill`](https://www.npmjs.com/package/promise-polyfill) and that has no protocol whatsoever for requests timing out. It will wait indefinitely for a response or until the connection drops.

**Comment by stevehobbsdev** [comment]
> Can you summarize what is the difference between the existing `getJSON` method and the newly introduced `fetchWithTimeout`? Both receive a timeout param and that confuses me.

@lbalmaceda `getJSON` _uses_ `fetchWithTimeout`. Right now on master, `getJSON` uses `fetch` directly but now uses this wrapped version that takes care of a timeout.

File: src/utils.ts (hunk 4 above)
Comment (lbalmaceda) [reply src/utils.ts]: Then, do we get a way to cancel that `fetch(url, options)` promise gracefully?

Comment (stevehobbsdev) [reply src/utils.ts]: Yes, I've done some research into this and we can cancel `fetch` through the use of [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController). However, it has [no support on IE11](https://caniuse.com/#feat=abortcontroller) and thus would require a [polyfill](https://www.npmjs.com/package/abortcontroller-polyfill) and would bump the size of the package.

Let me investigate how big that bump would be and come back to you.

Comment (stevehobbsdev) [reply src/utils.ts]: @lbalmaceda I've implemented `AbortController` now and added a polyfill for IE11.

The polyfill bumps the distributed package size up to 16.55kb, a 12.4% increase.

File: src/utils.ts [hunk 5]
```diff
@@ -189,14 +189,22 @@ export const bufferToBase64UrlEncoded = input => {
...
   return Promise.race([
-    fetch(url, options),
+    fetch(url, fetchOptions),
     new Promise((_, reject) => {
-      setTimeout(
-        () => reject(new Error("Timeout when executing 'fetch'")),
-        timeout
-      );
+      setTimeout(() => {
+        controller.abort();
```
Comment (lbalmaceda) [inline src/utils.ts]: Please make sure to test this call, I guess we could pass a mocked controller and asser the `abort` function got called in it?

File: src/utils.ts (hunk 4 above)
Comment (lbalmaceda) [reply src/utils.ts]: That doesn't seem like a huge increment, but I'd leave that to your discretion. Polyfill Q: Is there any way of only importing it on IE or is this something that if added, must be added in a general way?

Comment (stevehobbsdev) [reply src/utils.ts]: As it's all bundled, we have to add it in a generic way.

File: src/utils.ts (hunk 5 above)
Comment (stevehobbsdev) [reply src/utils.ts]: I don't think passing the AbortController through is the right call, but I have another idea. Stand by for an update on this tomorrow.

Comment (stevehobbsdev) [reply src/utils.ts]: @lbalmaceda I've added a way to test this, please re-review.

**Comment by stevehobbsdev** [comment]
@lbalmaceda thank you for being thorough!
